### PR TITLE
Allow printf-style exception messages

### PR DESCRIPTION
**Pull request by paradust7**
Add a constructor to BaseException to allow printf-style exceptions. For example:

```
if (size != expectedSize)
     throw MyException("Got size = %zu, expected %zu", size, expectedSize);
```
While technically not required, since one can always use a stringstream to construct a temporary string, I have an upcoming PR that benefits greater from being able to succinctly construct exception messages like this.

### CODE REVIEW

**Comment by paradust7** [comment]
When I made this PR, I hadn't noticed the existence of porting::mt_snprintf. I believe all uses of porting::mt_snprintf could be replaced with StringPrintf, and this would be cleaner in most cases, because there's no need for a temporary buffer at the call site.

The only thing mt_snprintf seems to do is use a special MSVC version of snprintf that supports positional arguments...but I don't see any case where positional arguments are used.

**Comment by sfan5** [comment]
> The only thing mt_snprintf seems to do is use a special MSVC version of snprintf that supports positional arguments...but I don't see any case where positional arguments are used.

It supports this because it's supposed to, that's not accidental. This is necessary because for translations you cannot assume the word order.

File: src/exceptions.h [hunk 1]
```diff
@@ -19,108 +19,84 @@ with this program; if not, write to the Free Software Foundation, Inc.,
...
+ * but GCC doesn't. So this macro is required for now.
+ */
+#define DEFINE_EXCEPTION_BASE(ClassName, EXTRA_CONSTRUCTORS) \
+	class ClassName : public BaseException { \
+	public: \
+		ClassName(const std::string &s) noexcept : BaseException(s) { } \
+		ClassName(std::string &&s) noexcept : BaseException(std::move(s)) { } \
+		PRINTF_ATTR(2, 3) \
+		explicit ClassName(PRINTF_FORMAT_PREFIX const char *format, ...) noexcept : \
+			BaseException(std::string()) \
```
Comment (sfan5) [inline src/exceptions.h:63]: can't this defer to BaseException with `std::forward`?

File: src/util/string.h [hunk 2]
```diff
@@ -761,3 +761,26 @@ std::string sanitizeDirName(const std::string &str, const std::string &optional_
...
+
+/**
+ *
+ * StringPrintf
+ *
+ * Create a std::string using a format string like printf.
+ *
+ */
+
+std::string StringPrintf(PRINTF_FORMAT_PREFIX const char *format, ...) PRINTF_ATTR(1, 2);
```
Comment (sfan5) [inline src/util/string.h:785]: should probably be called `string_printf` or maybe `strprintf`

File: src/nodedef.cpp [hunk 3]
```diff
@@ -654,22 +654,22 @@ void ContentFeatures::deSerialize(std::istream &is)
 
 		u8 tmp = readU8(is);
 		if (is.eof()) /* readU8 doesn't throw exceptions so we have to do this */
-			throw SerializationError("");
+			throw SerializationError("EOF");
```
Comment (sfan5) [inline src/nodedef.cpp:657]: these are immediately caught without going anywhere so need to fill anything in btw

File: src/exceptions.h (hunk 1 above)
Comment (paradust7) [reply src/exceptions.h:63]: Forwarding all arguments creates ambiguity with the other constructors.
Having the first arg be `const char* fmt` and forwarding the rest, breaks the printf formatting attribute.

File: src/util/string.h (hunk 2 above)
Comment (paradust7) [reply src/util/string.h:785]: `StringPrintf` is the name used in chromium (and android) for the same function. I would prefer this just out of habit, but if it is too inconsistent, I could replace it with `strprintf` (less typing anyway).

File: src/nodedef.cpp (hunk 3 above)
Comment (paradust7) [reply src/nodedef.cpp:657]: Strange enough, using "" as a format string generates a warning:

warning: zero-length gnu_printf format string [-Wformat-zero-length]

**Comment by paradust7** [comment]
I can't see any clean way of using MSVC's positional-arguments printf functions in StringPrintf, since there's no mechanism to query the size of buffer needed ahead of time. For now I've left translations still going through porting::mt_snprintf

I wonder if the positional argument thing is actually important. Unless I'm missing something, there are only three cases where it is used, and they all use the arguments in the same order anyway:

https://github.com/minetest/minetest/blob/76000e676bde0191d64dc0afd2f3ae74bb6c40b5/po/nn/minetest.po#L1964

https://github.com/minetest/minetest/blob/76000e676bde0191d64dc0afd2f3ae74bb6c40b5/po/th/minetest.po#L1990

https://github.com/minetest/minetest/blob/76000e676bde0191d64dc0afd2f3ae74bb6c40b5/po/pl/minetest.po#L1983

This one shows up on grep but looks like a typo:

https://github.com/minetest/minetest/blob/76000e676bde0191d64dc0afd2f3ae74bb6c40b5/po/eu/minetest.po#L744-L745

Are there other translations not in the repo?

File: src/exceptions.h (hunk 1 above)
Comment (paradust7) [reply src/exceptions.h:63]: Even doing direct delegation:

`using BaseException::BaseException;`

Breaks the printf attributes (but only on GCC). I would definitely consider this a GCC bug. I'll see about filing an issue.

Comment (paradust7) [reply src/exceptions.h:63]: Unfortunately it seems to be intentionally unsupported behavior:

https://github.com/gcc-mirror/gcc/blob/919822adc923b00e995e8a148bdb8115a794b47b/gcc/cp/call.cc#L9690

**Comment by paradust7** [comment]
The way fmtgettext is written right now, I don't think it would resize properly on Windows. The regular snprintf returns "The number of characters that would have been written if n had been sufficiently large, not counting the terminating null character.". But _vsprintf_p returns [the number of characters written, not including the terminating null character](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/vsprintf-p-vsprintf-p-l-vswprintf-p-vswprintf-p-l?view=msvc-170#return-value). This means that:

https://github.com/minetest/minetest/blob/9fc018ded10225589d2559d24a5db739e891fb31/src/gettext.h#L98-L104

will always be a no-op on Windows. The string would just be truncated to 256 bytes. To grow properly would require retrying until `len < buf.size() - 1`. It can certainly be done.

How well tested are these translations? I noticed just now, this seems like it would segfault because the original message has 1 argument, but the translation uses two:

https://github.com/minetest/minetest/blob/master/po/nn/minetest.po#L1958-L1968

https://github.com/minetest/minetest/blob/master/src/gui/guiConfirmRegistration.cpp#L90-L99

It's really ugly having so many format strings not checked by the compiler. It would be possible to replace gettext() with macros generated from the .po files (as a pre-compile step in cmake). The macros could resolve the translation at compile-time, and emit a switch with individual printf's for each locale's format string. I'm thinking something like:

```
auto str = LocalizedPrintf("Original string to be used for lookup %s", ...);
```

```
#define LocalizedPrintf(fmt, ...) []() { \
    switch (currentLocale()) { \
    case 'de': return StringPrintf(COMPILE_TIME_LOOKUP(fmt, 'de'), #__VA_ARGS__); \
    case 'es': return ...;
    ... \
    } \
}
```

Would you support such a thing? It's obviously a low priority but I could implement it eventually.

**Comment by paradust7** [comment]
I just realized, it should be fine to just call _vscprintf_p first, to get the buffer size.

The example from ffmpeg that mt_snprintf is based on:

https://github.com/FFmpeg/FFmpeg/blob/5ae9fa13f5ac640bec113120d540f70971aa635d/compat/msvcrt/snprintf.c#L65

Uses _vsnprintf, which returns -1 when the buffer overflows. But _vsprintf_p (used in mt_snprintf) does not act the same way, as far as I can tell.

**Comment by sfan5** [comment]
> How well tested are these translations? I noticed just now, this seems like it would segfault because the original message has 1 argument, but the translation uses two:

:shrug: The translation you found is marked "fuzzy" so gettext will skip it, but you're right that it would probably crash.

Don't worry about this though, guiConfirmationDialog is about to replaced with a Lua solution anyway.

> It would be possible to replace gettext() with macros generated from the .po files (as a pre-compile step in cmake)

That's an interesting idea, the single reason it'd not work out is that there's no way to ask gettext for the current langauge (really, we looked for one years ago but had to resort to a workaround).

**Comment by sfan5** [comment]
That's what we have, yes:
https://github.com/minetest/minetest/blob/9fc018ded10225589d2559d24a5db739e891fb31/src/network/clientpackethandler.cpp#L141-L144

**Comment by sfan5** [comment]
I have pretty much reviewed this already, here are my current points:
* I still think renaming the new helpers to snake case is a good idea
* what happened to using `_vsprintf_p` to support positional args on MSVC?
* can you split the commits so that changing all existing `mt_snprintf` is cleanly split from the rest?

**Pull request closed without merging**